### PR TITLE
ACS-2803: Add test to check if IN syntax works against ES

### PR DESCRIPTION
**Pull request by gevagel**

### CODE REVIEW

**Comment by montgolfiere** [comment]
I see this failed on Travis with 500 error. Anything obvious in logs (eg. if not on Travis then when run locally) otherwise needs to be debugged.

```
[ERROR] checkInSyntax(org.alfresco.elasticsearch.ElasticsearchCMISTests)  Time elapsed: 20.619 s  <<< FAILURE!
java.lang.AssertionError: Maximum retry period reached, test failed.
	at org.alfresco.elasticsearch.ElasticsearchCMISTests.checkInSyntax(ElasticsearchCMISTests.java:155)
Caused by: java.lang.AssertionError: Status code is not as expected. expected [200] but found [500]
	at org.alfresco.elasticsearch.ElasticsearchCMISTests.checkInSyntax(ElasticsearchCMISTests.java:155)
```

**Comment by montgolfiere** [comment]
@gevagel Have you managed to debug where (& why) this fails ?

**Comment by montgolfiere** [comment]
@gevagel What's the plan for this one ?

**Comment by montgolfiere** [comment]
Thanks @gevagel  - good to know.

Are you sure you really need custom content model ?

AFAIK, the tests via TestContainers allows for programmatic control of running specific tests with specific "feature flags" (test env config properties) such as with & without exact term matching, cc @tpage-alfresco 

- https://docs.alfresco.com/search-enterprise/latest/admin/#pre-indexing-considerations
- https://docs.alfresco.com/search-enterprise/latest/using/#search-in-fields

**Comment by montgolfiere** [comment]
See also: https://github.com/Alfresco/alfresco-enterprise-repo/pull/773

It would be good to review and approve both together. At minimum the happy path (ie. exact term enabled).

It might be that here we also go one step further (follow-on PR ?) and test with and without exact term (ie. -ve test) (eg. using current programmatic TestContainers approach).

File: tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java [hunk 1]
```diff
@@ -147,6 +147,14 @@ public void checkPermissionForUser2()
         searchQueryService.expectResultsFromQuery(query, user2, FILE_2_NAME, USER_2_FILE_NAME);
     }
 
+    @TestRail (description = "Check IN('value1','value2') syntax works.", section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION)
+    @Test (groups = TestGroup.SEARCH)
+    public void checkInSyntax()
+    {
+        SearchRequest query = req("cmis", "SELECT * FROM cmis:document WHERE cmis:name IN ('" + FILE_0_NAME + "') AND CONTAINS('*')");
```
[the remark this answers is not in the document]
Comment (montgolfiere) [reply tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java:162]: Make this consistent with https://github.com/Alfresco/alfresco-enterprise-repo/pull/773/files#diff-2466e672e08f8f55c9f5273d5f2a5555f148ce057aaf137da65bbd7df917f35cR57 (ie. check at least two IN values).

**Review by montgolfiere — CHANGES_REQUESTED** [review]
Added some initial comments. 

We should make this at least consistent with similar ent-repo test, ie. minimum, test with exact match enabled).

Also consider additional test scenarios (given separate test config, maybe separate test class) with alternative (arguably negative) behaviour, ie. exact match not enabled.

File: tests/tas-elasticsearch/src/test/resources/exactTermSearch.properties [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Exact Term search is switched off by default as it introduces index size overhead.
+#to enable it, please uncomment the following on a datatype or property name basis
+#alfresco.cross.locale.datatype.0={http://www.alfresco.org/model/dictionary/1.0}text
+#alfresco.cross.locale.datatype.1={http://www.alfresco.org/model/dictionary/1.0}content
+#alfresco.cross.locale.datatype.2={http://www.alfresco.org/model/dictionary/1.0}mltext
+#alfresco.cross.locale.property.0={http://www.alfresco.org/model/content/1.0}content
+alfresco.cross.locale.property.1={http://www.alfresco.org/model/content/1.0}name
```
Comment (montgolfiere) [inline tests/tas-elasticsearch/src/test/resources/exactTermSearch.properties:7]: Should this by `.0` (or does it not matter if only one property) ?

File: tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java [hunk 3]
```diff
@@ -147,6 +147,22 @@ public void checkPermissionForUser2()
         searchQueryService.expectResultsFromQuery(query, user2, FILE_2_NAME, USER_2_FILE_NAME);
     }
 
+    @TestRail (description = "Check that exact term search works. Needs exact term search to be enabled to pass.", section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION)
+    @Test (groups = TestGroup.SEARCH)
+    public void matchDocumentName()
+    {
+        SearchRequest query = req("cmis", "SELECT * FROM cmis:document WHERE cmis:name = '" + FILE_0_NAME + "' AND CONTAINS('*')");
```
Comment (montgolfiere) [inline tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java:154]: Do we really want the CONTAIN('*') here ? 

Wouldn't it be better to disable the option to go via DB (TMDQ) when starting repo ? At least for Elasticsearch CMIS Tests (and possibly AFTS ones as well) ?

Note: for now it requires the following property name (until we look to remove "solr." part):

`-Dsolr.query.cmis.queryConsistency=EVENTUAL`

And remove &/or refactor any tests that do not yet work, eg. basicCMISQueryAgainstDB(). See for example:

https://github.com/Alfresco/alfresco-enterprise-repo/pull/809/files#diff-2466e672e08f8f55c9f5273d5f2a5555f148ce057aaf137da65bbd7df917f35cR41-L44

File: tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java [hunk 4]
```diff
@@ -147,6 +147,22 @@ public void checkPermissionForUser2()
...
+        searchQueryService.expectResultsFromQuery(query, user1, FILE_0_NAME);
+    }
+
+    @TestRail (description = "Check IN('value1','value2') syntax works. Needs exact term search to be enabled to pass.", section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION)
+    @Test (groups = TestGroup.SEARCH)
+    public void checkInSyntax()
+    {
+        SearchRequest query = req("cmis", "SELECT * FROM cmis:document WHERE cmis:name IN ('" + FILE_0_NAME + "', '" + FILE_1_NAME + "') AND CONTAINS('*')");
```
Comment (montgolfiere) [inline tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java:162]: See comment above, re: CONTAINS('*')

**Review by montgolfiere — COMMENTED** [review]
Thanks @gevagel and @KikoUK 

I've added a few minor comments & possible improvements. 

For now, I suggest we try to be aligned and consistent with BasicQueryIT (in alfresco-enterprise-repo) ... at least where it makes sense.

**Review by KikoUK — APPROVED** [review]
LGTM subject to ✅ build and Jan's comments being addressed.

File: tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java [hunk 5]
```diff
@@ -147,6 +147,22 @@ public void checkPermissionForUser2()
         searchQueryService.expectResultsFromQuery(query, user2, FILE_2_NAME, USER_2_FILE_NAME);
     }
 
+    @TestRail (description = "Check that exact term search works. Needs exact term search to be enabled to pass.", section = TestGroup.SEARCH, executionType = ExecutionType.REGRESSION)
```
Comment (tpage-alfresco) [inline tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java:150]: I think we're really checking that we can match a document name (as the method name says). This uses exact term search, but I don't think an end user would think of it that way.

File: tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java (hunk 3 above)
Comment (tpage-alfresco) [reply tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java:154]: This sounds like a great improvement to these tests, but let's do it as a separate ticket.

Comment (montgolfiere) [reply tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchCMISTests.java:154]: Hopefully just a couple of lines (assuming no other fallout). But yes, it could also be done as part of ACS-2920 (if not sooner).

File: tests/tas-elasticsearch/src/test/resources/exactTermSearch.properties [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Exact Term search is switched off by default as it introduces index size overhead.
+#to enable it, please uncomment the following on a datatype or property name basis
+#alfresco.cross.locale.datatype.0={http://www.alfresco.org/model/dictionary/1.0}text
+#alfresco.cross.locale.datatype.1={http://www.alfresco.org/model/dictionary/1.0}content
+#alfresco.cross.locale.datatype.2={http://www.alfresco.org/model/dictionary/1.0}mltext
+#alfresco.cross.locale.property.0={http://www.alfresco.org/model/content/1.0}content
```
[the remark this answers is not in the document]
Comment (montgolfiere) [reply tests/tas-elasticsearch/src/test/resources/exactTermSearch.properties:6]: As per @tpage-alfresco we could remove commented lines.

Instead, we could also add a comment line which cross-references to similar test config for embedded IT tests running in alfresco-enterprise-repo.

https://github.com/Alfresco/alfresco-enterprise-repo/blob/master/repository/src/test/resources/alfresco/search/elasticsearch/config/exactTermSearch.properties

Interestingly that example has cm:content but not cm:name. Instead, it has d:text and d:mltext (presumably to cover all indexed text properties) ?

**Review by montgolfiere — APPROVED** [review]
See @tpage-alfresco comments.

Otherwise, I'm ok if we merge and then plan to improve and extend as part of additional sub-tasks, including fix for ACS-2920.
